### DISCUSSION TITLE
Is it possible to convert an AST back into the language it came from?

### DISCUSSION THREAD

**Discussion by mainrs** [tree-sitter/tree-sitter#1807] [Q&A] [answered]
I am trying to use tree-sitter to allow for structured inserts of data into source code files. Instead of saying "I want to add this text at line 25" I want to be able to say "Insert this line after the last import statement".

Parsing and editing the AST is no problem. But it seems that converting it back to a string is not possible using tree-sitter. What approaches can I take?

**Comment by ahelwer** [comment, marked as answer]
Why do you care about converting it back to a string when you still have the string? You can certainly use the AST to find the correct edit position in the string, make changes to the string, then do an incremental reparse to update the AST.

**Reply by mainrs** [reply]
I didn't think about that to be honest! Thanks for mentioning it! Was an oversight be me!

**Reply by danielo515** [reply]
So does the parsed AST contain information about the location of the nodes?
In any case, I think generating code from an AST is not a trivial task, and just string processing is generally not enough.

**Reply by ahelwer** [reply]
@danielo515 yes it does contain info about locations of nodes.
